runtime/lua/vim/ui/img.lua

```diff
@@ -0,0 +1,557 @@
+M.providers = require('vim.ui.img.providers')
+
+---Creates a new image instance, optionally taking pre-loaded bytes.
+---@param opts string|{bytes?:string, filename:string}
+---@return vim.ui.Image
+function M.new(opts)
+  vim.validate('opts', opts, { 'string', 'table' })
+  if type(opts) == 'table' then
+    vim.validate('opts.bytes', opts.bytes, 'string', true)
+    vim.validate('opts.filename', opts.filename, 'string')
+  end
+
+  local instance = {}
+  setmetatable(instance, M)
+
+  instance.id = LAST_IMAGE_ID + 1
+  instance.__metadata = {}
+  if type(opts) == 'table' then
+    instance.bytes = opts.bytes
+    instance.filename = opts.filename
+  elseif type(opts) == 'string' then
+    instance.filename = opts
+  end
+
+  -- Bump our counter for future image ids
+  LAST_IMAGE_ID = instance.id
+
+  return instance
+end
+
+---Loads bytes for an image from a local file.
+---@param filename string
+---@return vim.ui.img.utils.Promise<vim.ui.Image>
+function M.load(filename)
+  local promise = require('vim.ui.img.utils.promise').new({
+    context = 'image.load',
+  })
+
+  local img = M.new({ filename = filename })
+  img
+    :reload()
+    :on_ok(function()
+      promise:ok(img)
+    end)
+    :on_fail(function(err)
+      promise:fail(err)
+    end)
+
+  return promise
+end
+
+---Reloads the bytes for an image from its filename.
+---@return vim.ui.img.utils.Promise<vim.NIL>
+function M:reload()
+  local filename = self.filename
+  local promise = require('vim.ui.img.utils.promise').new({
+    context = 'image.reload',
+  })
+
+  ---@param err string|nil
+  ---@return boolean
+  local function report_err(err)
+    if err then
+      promise:fail(err)
+    end
+
+    return err ~= nil
+  end
+
+  vim.uv.fs_stat(filename, function(stat_err, stat)
+    if report_err(stat_err) then
+      return
+    end
+    if not stat then
+      report_err('missing stat')
+      return
+    end
+
+    vim.uv.fs_open(filename, 'r', 644, function(open_err, fd)
+      if report_err(open_err) then
+        return
+      end
+      if not fd then
+        report_err('missing fd')
+        return
+      end
+
+      vim.uv.fs_read(fd, stat.size, -1, function(read_err, bytes)
+        if report_err(read_err) then
+          return
+        end
+
+        self.bytes = bytes or ''
+        self.filename = filename
+
+        vim.uv.fs_close(fd, function()
+          promise:ok(vim.NIL)
+        end)
+      end)
+    end)
+  end)
+
+  return promise
+end
+
+---Returns the byte length of the image's bytes, or 0 if not loaded.
+---@return integer
+function M:len()
+  return string.len(self.bytes or '')
+end
+
+---Returns a hash (sha256) of the image's bytes.
+---@return string
+function M:hash()
+  return vim.fn.sha256(self.bytes or '')
+end
+
+---Check if the image is PNG format, optionally loading the magic number of the image.
+---Will throw an error if unable to load the bytes of the file.
+---Works without ImageMagick.
+---@return boolean
+function M:is_png()
```

Comment:
strange for a boolean getter to throw. can this be `format():'png'|nil` instead, where `nil` means "unknown"? or maybe `format():'png'|'error'|nil`

Reply:
Yeah, can do that instead. Do you expect to support other formats other than PNG down the line? If so, one of the `format` functions makes sense. If not, the `is_png` makes more sense.